### PR TITLE
Allow mapping invalid leaf entries

### PR DESCRIPTION
**Pull request by jakobvukalovic**
Break up invalid block entries into a new subtable when a new range smaller than the block is mapped. Also, allow mapping block and page entries without a VALID bit.

### CODE REVIEW

File: src/paging.rs [hunk 1]
```diff
@@ -653,12 +661,8 @@ pub struct Descriptor(usize);
 impl Descriptor {
     const PHYSICAL_ADDRESS_BITMASK: usize = !(PAGE_SIZE - 1) & !(0xffff << 48);
 
-    fn output_address(self) -> Option<PhysicalAddress> {
-        if self.is_valid() {
-            Some(PhysicalAddress(self.0 & Self::PHYSICAL_ADDRESS_BITMASK))
-        } else {
-            None
-        }
+    fn output_address(self) -> PhysicalAddress {
```
Comment (ptosi) [inline src/paging.rs:664]: Should this be named `physical_address`, for clarity? Or even be `impl From<Descriptor> for PhysicalAddress` (not sure why it wasn't already `TryFrom`)?

File: src/paging.rs [hunk 2]
```diff
@@ -708,8 +711,10 @@ impl Descriptor {
 impl Debug for Descriptor {
     fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
         write!(f, "{:#016x}", self.0)?;
-        if let (Some(flags), Some(address)) = (self.flags(), self.output_address()) {
-            write!(f, " ({}, {:?})", address, flags)?;
+        if self.is_valid() {
+            if let Some(flags) = self.flags() {
+                write!(f, " ({}, {:?})", self.output_address(), flags)?;
```
Comment (ptosi) [inline src/paging.rs:716]: nit: format the address?

File: src/paging.rs [hunk 3]
```diff
@@ -835,14 +840,14 @@ mod tests {
...
-            Attributes::TABLE_OR_PAGE | Attributes::USER | Attributes::SWFLAG_1,
+            Attributes::TABLE_OR_PAGE | Attributes::USER | Attributes::SWFLAG_1 | Attributes::VALID,
         );
         assert!(desc.is_valid());
         assert_eq!(
             desc.flags().unwrap(),
             Attributes::TABLE_OR_PAGE | Attributes::USER | Attributes::SWFLAG_1 | Attributes::VALID
         );
-        assert_eq!(desc.output_address().unwrap(), PhysicalAddress(0x12340000));
+        assert_eq!(desc.output_address(), PhysicalAddress(0x12340000));
```
Comment (ptosi) [inline src/paging.rs:850]: nit: make `0x12340000` a local `const`?

**Review by ptosi — CHANGES_REQUESTED** [review]

File: src/paging.rs [hunk 4]
```diff
@@ -506,19 +506,22 @@ impl<T: Translation> PageTableWithLevel<T> {
...
-                        subtable.map_range(
-                            translation,
-                            &MemoryRegion::new(a, b),
-                            old_pa,
-                            old_flags,
-                        );
+                    if let Some(old_flags) = old.flags() {
+                        if !old_flags.contains(Attributes::TABLE_OR_PAGE) {
+                            let old_pa = old.output_address();
+                            // Old was a valid block entry, so we need to split it.
```
Comment (qwandor) [inline src/paging.rs:512]: How do you know it was valid?

File: src/paging.rs [hunk 5]
```diff
@@ -545,12 +549,16 @@ impl<T: Translation> PageTableWithLevel<T> {
                     i += 1;
                 }
                 if i - 1 == first_zero {
-                    writeln!(f, "{:indentation$}{}: 0", "", first_zero)?;
+                    writeln!(f, "{:indentation$}{: <WIDTH$}: 0", "", first_zero)?;
                 } else {
-                    writeln!(f, "{:indentation$}{}-{}: 0", "", first_zero, i - 1)?;
+                    writeln!(f, "{:indentation$}{: <WIDTH$}-{}: 0", "", first_zero, i - 1,)?;
```
Comment (qwandor) [inline src/paging.rs:554]: Why the extra comma at the end here?

File: src/paging.rs [hunk 6]
```diff
@@ -545,12 +549,16 @@ impl<T: Translation> PageTableWithLevel<T> {
                     i += 1;
                 }
                 if i - 1 == first_zero {
-                    writeln!(f, "{:indentation$}{}: 0", "", first_zero)?;
+                    writeln!(f, "{:indentation$}{: <WIDTH$}: 0", "", first_zero)?;
```
Comment (qwandor) [inline src/paging.rs:552]: This change looks unrelated, maybe it should be a separate commit.

File: src/paging.rs (hunk 1 above)
Comment (qwandor) [reply src/paging.rs:664]: I think the current name is clearer; it's not just any physical address, it's specifically the output address of the page table mapping.

File: src/paging.rs (hunk 2 above)
Comment (qwandor) [reply src/paging.rs:716]: How? The `Debug` implementation for `PhysicalAddress` should already do the right thing.

**Review by qwandor — COMMENTED** [review]
This is very much a breaking change, so please note it in CHANGELOG.md as such. You also need to update the Rustdoc for the various `map_range` methods as they no longer necessarily map the range, and update the examples.

File: src/paging.rs (hunk 3 above)
Comment (jakobvukalovic) [reply src/paging.rs]: Updated.

File: src/paging.rs (hunk 5 above)
Comment (jakobvukalovic) [reply src/paging.rs]: Typo, removed.

File: src/paging.rs (hunk 4 above)
Comment (jakobvukalovic) [reply src/paging.rs]: Updated, the block no longer needs to have the valid bit.

File: src/paging.rs (hunk 6 above)
Comment (jakobvukalovic) [reply src/paging.rs:554]: Moved to a new commit.

File: src/paging.rs [hunk 7]
```diff
@@ -506,19 +506,22 @@ impl<T: Translation> PageTableWithLevel<T> {
...
+                                old_pa,
+                                old_flags,
+                            );
+                        }
                     }
-                    entry.set(subtable_pa, Attributes::TABLE_OR_PAGE);
+                    entry.set(subtable_pa, Attributes::TABLE_OR_PAGE | Attributes::VALID);
                     subtable
                 };
                 subtable.map_range(translation, &chunk, pa, flags);
```
[the remark this answers is not in the document]
Comment (jakobvukalovic) [reply src/paging.rs:529]: I have updated the docs and changelog as suggested [here](https://github.com/google/aarch64-paging/pull/27#pullrequestreview-1385790380).

**Comment by jakobvukalovic** [comment]
> This is very much a breaking change, so please note it in CHANGELOG.md as such. You also need to update the Rustdoc for the various `map_range` methods as they no longer necessarily map the range, and update the examples.

I have now updated the changelog, docs, and examples.

**Review by qwandor — APPROVED** [review]
